src/kdbindings/connection_evaluator.h

```diff
@@ -0,0 +1,84 @@
+#pragma once
+
+#include <functional>
+#include <list>
+#include <mutex>
+
+#include <kdbindings/connection_handle.h>
+
+namespace KDBindings {
+
+/**
+ * @brief Manages and evaluates deferred Signal connections.
+ *
+ * The ConnectionEvaluator class is responsible for managing and evaluating connections
+ * to Signals. It provides mechanisms to delay and control the evaluation of connections.
+ * It therefore allows controlling when and on which thread slots connected to a Signal are executed.
+ *
+ * @see Signal::connectDeferred()
+ */
+class ConnectionEvaluator
+{
+
+public:
+    ConnectionEvaluator() = default;
+
+    // ConnectionEvaluator is not copyable, as it is designed to manage connections,
+    // and copying it could lead to unexpected behavior, including duplication of connections and issues
+    // related to connection lifetimes. Therefore, it is intentionally made non-copyable.
+    ConnectionEvaluator(const ConnectionEvaluator &) noexcept = delete;
+
+    ConnectionEvaluator &operator=(const ConnectionEvaluator &) noexcept = delete;
+
+    // ConnectionEvaluators are not moveable, as they are captures by-reference
+    // by the Signal, so moving them would lead to a dangling reference.
+    ConnectionEvaluator(ConnectionEvaluator &&other) noexcept = delete;
+
+    ConnectionEvaluator &operator=(ConnectionEvaluator &&other) noexcept = delete;
+
+    /**
+     * @brief Evaluate the deferred connections.
+     *
+     * This function is responsible for evaluating and executing deferred connections.
+     * And this function ensures thread safety
+     */
+    void evaluateDeferredConnections()
+    {
+        std::vector<std::pair<const ConnectionHandle *, std::function<void()>>> movedConnections;
+
+        {
+            std::lock_guard<std::mutex> lock(m_connectionsMutex);
+
+            // Move out the current connections and replace the original vector with an empty one.
+            movedConnections = std::move(m_deferredConnections);
```

Comment:
I still like this trick, as we avoid locking the mutex for long.
But it's unfortunately incompatible with the approach of removing queued slot invocations on disconnect.

As in a concurrent context the following is possible:
- A deferred signal is emitted
- evaluateDeferredConnections is called and the signal is copied, but the following loop isn't run yet.
- `disconnect` is called, but finds no queued invocations, as those have already been cleared
- `movedConnections` is iterated and the slot is still invoked, even after `disconnect` returns.

So we need to either remove this optimization :crying_cat_face: or revert back to leaving disconnected emit's queued.

We should really try this out in a larger context, to see if the resource management issues associated with evaluating already disconnected invocations is worth it, or even preferable.